### ISSUE TITLE
New interpreter coerces fields that are not part of the query

### ISSUE THREAD

**Issue by vfonic** [rmosolgo/graphql-ruby#2788]
I've turned on the new interpreter and now I'm receiving errors about `DateTime` type not being able to coerce `ActiveSupport::TimeWithZone`:

```
     TypeError:
       no implicit conversion of ActiveSupport::TimeWithZone into String
     # .../graphql-1.10.3/lib/graphql/types/iso_8601_date_time.rb:44:in `match'
     # .../graphql-1.10.3/lib/graphql/types/iso_8601_date_time.rb:44:in `iso8601'
     # .../graphql-1.10.3/lib/graphql/types/iso_8601_date_time.rb:44:in `coerce_input'
     # .../graphql-1.10.3/lib/graphql/schema/member/has_arguments.rb:104:in `block (2 levels) in coerce_arguments'
     # .../graphql-1.10.3/lib/graphql/execution/errors.rb:30:in `with_error_handling'
     # .../graphql-1.10.3/lib/graphql/schema/member/has_arguments.rb:99:in `block in coerce_arguments'
     # .../graphql-1.10.3/lib/graphql/schema/member/has_arguments.rb:74:in `each'
     # .../graphql-1.10.3/lib/graphql/schema/member/has_arguments.rb:74:in `coerce_arguments'
     # .../graphql-1.10.3/lib/graphql/execution/interpreter/runtime.rb:480:in `arguments'
     # .../graphql-1.10.3/lib/graphql/execution/interpreter/runtime.rb:171:in `block in evaluate_selections'
     # .../graphql-1.10.3/lib/graphql/execution/interpreter/runtime.rb:124:in `each'
     # .../graphql-1.10.3/lib/graphql/execution/interpreter/runtime.rb:124:in `evaluate_selections'
     # .../graphql-1.10.3/lib/graphql/execution/interpreter/runtime.rb:323:in `block in continue_field'
     # .../graphql-1.10.3/lib/graphql/execution/interpreter/runtime.rb:435:in `after_lazy'
     # .../graphql-1.10.3/lib/graphql/execution/interpreter/runtime.rb:318:in `continue_field'
     # .../graphql-1.10.3/lib/graphql/execution/interpreter/runtime.rb:343:in `block (2 levels) in continue_field'
     # .../graphql-1.10.3/lib/graphql/execution/interpreter/runtime.rb:435:in `after_lazy'
     # .../graphql-1.10.3/lib/graphql/execution/interpreter/runtime.rb:340:in `block in continue_field'
     # .../graphql-1.10.3/lib/graphql/execution/interpreter/runtime.rb:333:in `each'
     # .../graphql-1.10.3/lib/graphql/execution/interpreter/runtime.rb:333:in `continue_field'
     # .../graphql-1.10.3/lib/graphql/execution/interpreter/runtime.rb:228:in `block (3 levels) in evaluate_selections'
     # .../graphql-1.10.3/lib/graphql/execution/interpreter/runtime.rb:435:in `after_lazy'
     # .../graphql-1.10.3/lib/graphql/execution/interpreter/runtime.rb:222:in `block (2 levels) in evaluate_selections'
     # .../graphql-1.10.3/lib/graphql/execution/interpreter/runtime.rb:359:in `block in resolve_with_directives'
     # .../graphql-1.10.3/lib/graphql/execution/interpreter/runtime.rb:365:in `run_directive'
     # .../graphql-1.10.3/lib/graphql/execution/interpreter/runtime.rb:359:in `resolve_with_directives'
     # .../graphql-1.10.3/lib/graphql/execution/interpreter/runtime.rb:211:in `block in evaluate_selections'
     # .../graphql-1.10.3/lib/graphql/execution/interpreter/runtime.rb:124:in `each'
     # .../graphql-1.10.3/lib/graphql/execution/interpreter/runtime.rb:124:in `evaluate_selections'
     # .../graphql-1.10.3/lib/graphql/execution/interpreter/runtime.rb:323:in `block in continue_field'
     # .../graphql-1.10.3/lib/graphql/execution/interpreter/runtime.rb:435:in `after_lazy'
     # .../graphql-1.10.3/lib/graphql/execution/interpreter/runtime.rb:318:in `continue_field'
     # .../graphql-1.10.3/lib/graphql/execution/interpreter/runtime.rb:352:in `continue_field'
     # .../graphql-1.10.3/lib/graphql/execution/interpreter/runtime.rb:228:in `block (3 levels) in evaluate_selections'
     # .../graphql-1.10.3/lib/graphql/execution/interpreter/runtime.rb:435:in `after_lazy'
     # .../graphql-1.10.3/lib/graphql/execution/interpreter/runtime.rb:222:in `block (2 levels) in evaluate_selections'
     # .../graphql-1.10.3/lib/graphql/execution/interpreter/runtime.rb:359:in `block in resolve_with_directives'
     # .../graphql-1.10.3/lib/graphql/execution/interpreter/runtime.rb:365:in `run_directive'
     # .../graphql-1.10.3/lib/graphql/execution/interpreter/runtime.rb:359:in `resolve_with_directives'
     # .../graphql-1.10.3/lib/graphql/execution/interpreter/runtime.rb:211:in `block in evaluate_selections'
     # .../graphql-1.10.3/lib/graphql/execution/interpreter/runtime.rb:124:in `each'
     # .../graphql-1.10.3/lib/graphql/execution/interpreter/runtime.rb:124:in `evaluate_selections'
     # .../graphql-1.10.3/lib/graphql/execution/interpreter/runtime.rb:60:in `run_eager'
     # .../graphql-1.10.3/lib/graphql/execution/interpreter.rb:70:in `block in evaluate'
     # .../graphql-1.10.3/lib/graphql/tracing.rb:66:in `block in trace'
     # .../graphql-1.10.3/lib/graphql/tracing.rb:82:in `block (2 levels) in call_tracers'
     # .../graphql-1.10.3/lib/graphql/tracing.rb:80:in `call_tracers'
     # .../graphql-1.10.3/lib/graphql/tracing.rb:82:in `block in call_tracers'
     # .../graphql-1.10.3/lib/graphql/schema/timeout.rb:84:in `trace'
     # .../graphql-1.10.3/lib/graphql/tracing.rb:82:in `call_tracers'
     # .../graphql-1.10.3/lib/graphql/tracing.rb:66:in `trace'
     # .../graphql-1.10.3/lib/graphql/execution/interpreter.rb:69:in `evaluate'
     # .../graphql-1.10.3/lib/graphql/execution/interpreter.rb:42:in `begin_query'
     # .../graphql-1.10.3/lib/graphql/execution/multiplex.rb:113:in `begin_query'
     # .../graphql-1.10.3/lib/graphql/execution/multiplex.rb:84:in `block in run_as_multiplex'
     # .../graphql-1.10.3/lib/graphql/execution/multiplex.rb:83:in `map'
     # .../graphql-1.10.3/lib/graphql/execution/multiplex.rb:83:in `run_as_multiplex'
     # .../graphql-1.10.3/lib/graphql/execution/multiplex.rb:62:in `block (2 levels) in run_queries'
     # .../graphql-1.10.3/lib/graphql/execution/multiplex.rb:196:in `block in instrument_and_analyze'
     # .../graphql-1.10.3/lib/graphql/execution/instrumentation.rb:29:in `block (2 levels) in apply_instrumenters'
     # .../graphql-1.10.3/lib/graphql/execution/instrumentation.rb:46:in `block (2 levels) in each_query_call_hooks'
     # .../graphql-1.10.3/lib/graphql/execution/instrumentation.rb:41:in `each_query_call_hooks'
     # .../graphql-1.10.3/lib/graphql/execution/instrumentation.rb:45:in `block in each_query_call_hooks'
     # .../graphql-1.10.3/lib/graphql/execution/instrumentation.rb:72:in `call_hooks'
     # .../graphql-1.10.3/lib/graphql/execution/instrumentation.rb:44:in `each_query_call_hooks'
     # .../graphql-1.10.3/lib/graphql/execution/instrumentation.rb:27:in `block in apply_instrumenters'
     # .../graphql-1.10.3/lib/graphql/execution/instrumentation.rb:72:in `call_hooks'
     # .../graphql-1.10.3/lib/graphql/execution/instrumentation.rb:26:in `apply_instrumenters'
     # .../graphql-1.10.3/lib/graphql/execution/multiplex.rb:175:in `instrument_and_analyze'
     # .../graphql-1.10.3/lib/graphql/execution/multiplex.rb:61:in `block in run_queries'
     # .../graphql-1.10.3/lib/graphql/tracing.rb:66:in `block in trace'
     # .../graphql-1.10.3/lib/graphql/tracing.rb:82:in `block (2 levels) in call_tracers'
     # .../graphql-1.10.3/lib/graphql/tracing.rb:80:in `call_tracers'
     # .../graphql-1.10.3/lib/graphql/tracing.rb:82:in `block in call_tracers'
     # .../graphql-1.10.3/lib/graphql/schema/timeout.rb:60:in `trace'
     # .../graphql-1.10.3/lib/graphql/tracing.rb:82:in `call_tracers'
     # .../graphql-1.10.3/lib/graphql/tracing.rb:66:in `trace'
     # .../graphql-1.10.3/lib/graphql/execution/multiplex.rb:59:in `run_queries'
     # .../graphql-1.10.3/lib/graphql/execution/multiplex.rb:49:in `run_all'
     # .../graphql-1.10.3/lib/graphql/schema.rb:1557:in `multiplex'
     # .../graphql-1.10.3/lib/graphql/schema.rb:1528:in `execute'
```

The query doesn't contain any fields that have `DateTimeType`, but there are fields that have `DateTimeType` within the type that's being fetched.

Oddly enough, this only happens on "nested" query like this one:

```graphql
query {
  communities {
    liveState
    displayName
    events {
      nodes {
        title
        publishState
      }
    }
  }
}
```

It doesn't happen on these queries:

```graphql
query {
  communities {
    liveState
    displayName
  }
}
```

```graphql
query($slug: String!) {
  event(slug: $slug) {
    title
    publishState
  }
}
```

**Comment by swalkinshaw** [comment]
Does the `events` fields have any optional arguments? Or arguments with default values? (I'm just guessing `events` but it could be another field).
 
The error is coming from argument coercion, not field. It looks like the interpreter "evaluates" arguments defined on fields even if none are specified - which should be correct in case they have default values.

**Comment by vfonic** [comment]
Wow! Thanks for such prompt replies! 👍 🚀 

> Does the `events` fields have any optional arguments? Or arguments with default values? (I'm just guessing `events` but it could be another field).

You guessed it right! :) 

I have a resolver on `EventType`. Here's how it all looks together:

```rb
module Types
  class CommunityType < BaseObject
    description 'Get Community'

    field :events, EventType.connection_type, null: false, resolver: Resolvers::EventResolver
```

```rb
require 'search_object'
require 'search_object/plugin/graphql'

module Resolvers
  class EventResolver
    include SearchObject.module(:graphql)
    include RecordSearch

    scope { self.object }

    # here's the default value!
    option(:from_date, type: Types::DateTimeType, default: Time.zone.now.beginning_of_day) do |scope, value|
      if scope.respond_to?(:subscribed_events)
        scope.subscribed_events.after(value)
      else
        scope.published_events.after(value) + scope.published_child_communities_events.after(value)
      end
    end
  end
end
```

It's interesting that it used to work before with the old interpreter.

Here's the `schema.graphql` dump excerpt:

```
  events(
    """
    Returns the elements in the list that come after the specified cursor.
    """
    after: String

    """
    Returns the elements in the list that come before the specified cursor.
    """
    before: String

    """
    Returns the first _n_ elements from the list.
    """
    first: Int
    fromDate: ISO8601DateTime = "2020-02-28T00:00:00Z"

    """
    Returns the last _n_ elements from the list.
    """
    last: Int
    search: String
  ): EventConnection!
```

I never use this default value, but I remember I had to set it as otherwise I was receiving some errors. I believe the issue was that the default value is hardcoded to when the schema dump is run, which makes sense. But what I needed was: if user doesn't provide the value, use the current time aka "now", so I do that programatically.

I guess the solution is to convert the `TimeWithZone` object to iso date so that interpreter can coerce it correctly? Let me know if you know any better way to achieve any of this all.

Thanks!

**Comment by rmosolgo** [comment]
> the solution is to convert

Interestingly, the code that raised an error here expects a _string_: 

https://github.com/rmosolgo/graphql-ruby/blob/85512cfee6d8d11bfd7a8fd6fad2d1a0c9acf896/lib/graphql/types/iso_8601_date_time.rb#L40-L47

I would guess that the difference is that, previously, GraphQL expected an "application object" as a default value -- a `DateTime` of some kind. But I must have changed it so that it's expecting a "graphql value" -- a stringified datetime. 

I think this is a bug -- I didn't mean to change that, but I was just recently messing with how inputs are coerced.

**Comment by swalkinshaw** [comment]
> I never use this default value, but I remember I had to set it as otherwise I was receiving some errors. I believe the issue was that the default value is hardcoded to when the schema dump is run, which makes sense. But what I needed was: if user doesn't provide the value, use the current time aka "now", so I do that programatically.

Regardless of the potential bug, I'd probably change this behaviour.

It's strange to hardcoded a default value for a timestamp when you want `Time.now`. The argument is already optional so it seems like the resolver should just apply the default behaviour if no input was specified?

`EventResolver`  should basically do `from_date || Time.zone.now.beginning_of_day` within the actual field resolver method.

**Comment by vfonic** [comment]
@swalkinshaw you're right! I don't remember why I had to add that field's default value, but I'll see if I can simply remove it. Thanks for the advice!

Regarding the actual bug reported, I guess that's on @rmosolgo to decide. I'm thinking maybe having the interpreter break could be the desired behavior to early catch any kind of potential issues this might cause.

**Comment by rmosolgo** [comment]
Although the issue described here was a legit bug, it hasn't bothered anyone else, and by this time, people have probably made the necessary adjustments for the current behavior. So, I don't plan to make any other changes here.
